chainstate/src/detail/tests/syncing_tests.rs

```diff
@@ -187,8 +203,8 @@ fn get_headers_different_chains() {
             );
         }
 
-        btf1.create_chain(&prev.id, rng.gen_range(32..256)).unwrap();
-        btf2.create_chain(&prev.id, rng.gen_range(256..512)).unwrap();
+        btf1.create_chain(&prev.id, rng.gen_range(32..256), &mut rng).unwrap();
+        btf2.create_chain(&prev.id, rng.gen_range(256..512), &mut rng).unwrap();
```

Comment:
I am wondering if we can improve the case where the `rng` is used in multiple places.

Let me outline the issue first. Let's say this test fails with a known seed. You want to try if the failure still reproduces with smaller chains since debugging the processing of a chain with hundreds of blocks may be a bit inconvenient. So you adjust the range of the first chain length to `2..4`. Now the first range is shorter so after generating the first chain, the rng is in a different state than it was when the test ran initially. That implies that the second chain will get different length and contents.

If instead we did something like this ugly pseudocode (could be made nicer with helper methods etc):
```rust
let mut chain1_rng = make_seedable_rng(rng.gen_seed());
let mut chain2_rng = make_seedable_rng(rng.gen_seed());
btf1.create_chain(&prev.id, rng.gen_range(32..256), &mut chain1_rng).unwrap();
btf2.create_chain(&prev.id, rng.gen_range(256..512), &mut chain2_rng).unwrap();
```
This way, tinkering with length of the first chain does not change the length or contents of the second chain which is better for reproducibility. The interface can be made more conducive to this kind of approach if the `create_chain` (and similar) method took the rng by value (instead of `&mut`), so it is no longer accessible after the call, so its final state does not influence what is generated after it.

Possible drawbacks are more implementation effort. Also I am not sure if this "forking" behaviour would yield statistically robust results with some weaker PRNGs. Should be fine with ChaCha though.

Thoughts / comments?

Reply:
I don't really care... if you guys feel this will help, please go ahead and do it.

Me, when I fall into this issue of lots of stuff to be done to reproduce a problem, usually I follow a few steps to shrink the problem as much as possible. Given that I have the seed that originally created the problem, I can always fall back to that. But what you can do usually, is reduce the chains lengths, and rerun the tests virtually infinitely many times until the problem happens again. Usually most of the errors show up easily if there's a logic error. If it's a rare error (which is rare to happen to us, and rare to reproduce), I just use the original seed and live with it.

Again, I don't care... if you guys feel you wanna do some gymnastics with multiple seeds, I'm fine with that as long as everything is reproducible... I just think it's an overkill. At least for now.